src/py/flwr/server/state/state_test.py

```diff
@@ -0,0 +1,421 @@
+
+    def test_task_ins_store_anonymous_and_fail_retrieving_identitiy(self) -> None:
+        """Store anonymous TaskIns and fail to retrieve it."""
+        # Prepare
+        state: State = self.state_factory()
+        task_ins = create_task_ins(consumer_node_id=0, anonymous=True)
+
+        # Execute
+        _ = state.store_task_ins(task_ins)
+        task_ins_list = state.get_task_ins(node_id=1, limit=None)
+
+        # Assert
+        assert len(task_ins_list) == 0
+
+    def test_task_ins_store_identity_and_fail_retrieving_anonymous(self) -> None:
+        """Store identity TaskIns and fail retrieving it as anonymous."""
+        # Prepare
+        state: State = self.state_factory()
+        task_ins = create_task_ins(consumer_node_id=1, anonymous=False)
+
+        # Execute
+        _ = state.store_task_ins(task_ins)
+        task_ins_list = state.get_task_ins(node_id=None, limit=None)
+
+        # Assert
+        assert len(task_ins_list) == 0
+
+    def test_task_ins_store_identity_and_retrieve_identity(self) -> None:
+        """Store identity TaskIns and retrieve it."""
+        # Prepare
+        state: State = self.state_factory()
+        task_ins = create_task_ins(consumer_node_id=1, anonymous=False)
+
+        # Execute
+        task_ins_uuid = state.store_task_ins(task_ins)
+        task_ins_list = state.get_task_ins(node_id=1, limit=None)
+
+        # Assert
+        assert len(task_ins_list) == 1
+
+        retrieved_task_ins = task_ins_list[0]
+        assert retrieved_task_ins.task_id == str(task_ins_uuid)
+
+    def test_task_ins_store_delivered_and_fail_retrieving(self) -> None:
+        """Fail retrieving delivered task."""
+        # Prepare
+        state: State = self.state_factory()
+        task_ins = create_task_ins(consumer_node_id=1, anonymous=False)
+
+        # Execute
+        _ = state.store_task_ins(task_ins)
+
+        # Get once to set delivered
+        task_ins_list = state.get_task_ins(node_id=1, limit=None)
+
+        assert len(task_ins_list) == 1
+
+        # Get twice to get nothing
+        task_ins_list = state.get_task_ins(node_id=1, limit=None)
+
+        # Assert
+        assert len(task_ins_list) == 0
+
+    def test_get_task_ins_limit_throws_for_limit_zero(self) -> None:
+        """Fail call with limit=0."""
+        # Prepare
+        state: State = self.state_factory()
+
+        # Execute & Assert
+        with self.assertRaises(AssertionError):
+            state.get_task_ins(node_id=1, limit=0)
+
+    # TaskRes tests
+    def test_task_res_store_and_retrieve_by_task_ins_id(self) -> None:
+        """Store TaskRes retrieve it by task_ins_id."""
+        # Prepare
+        state: State = self.state_factory()
+        task_ins_id = uuid4()
+        task_res = create_task_res(
+            producer_node_id=0, anonymous=True, ancestry=[str(task_ins_id)]
+        )
+
+        # Execute
+        task_res_uuid = state.store_task_res(task_res)
```

Comment:
Side note: storing a `TaskRes` that references a non-existent `TaskIns` should probably result in an error. It's ok to leave it like that for now though.